ARC-0004: Misc changes (#55)
algorandfoundation/ARCs: ARCs/arc-0004.md

> Comments like this are non-normative.

> All strings of type `string` are UTF-8 encoded.

Author: I was thinking the `string` in the TS interfaces. But what about the strings in the ABI?

> All strings of type `string` are UTF-8 encoded.

Reviewer: Got it, maybe this would be more clear?

```suggestion
Interfaces are defined in TypeScript. All the objects that are defined are valid JSON objects, and all JSON `string` types are UTF-8 encoded.
```

And ABI strings are "assumed to contain UTF-8 encoded content."

#### Method Signature

A method signature is a unique identifier for a method. The signature
is a string that consists of the method's name, an open parenthesis, a
comma-separated list of the types of its arguments, a closing
parenthesis, and the method's return type, or `void` if it does not

> return a value. The names of arguments **MUST NOT** included in a

Reviewer 2: Must not be

interface Interface {
  /** A user-friendly name for the interface */
  name: string,
  /** Optional, user-friendly description for the interface */
  desc?: string,
  /** All of the methods that the interface contains */
  methods: Method[]
}
```

Interface names **MUST** satisfy the regular expression `[_A-Za-z][A-Za-z0-9_]*`.
Interface names starting by `ARC` are reserved to interfaces defined in ARC.
Interfaces defined in `ARC-XXXX` (where `XXXX` is a 0-padded number) **SHOULD**

> start with `ARC-XXXX`.

Reviewer: ```suggestion
start with `ARC_XXXX`.
```
The hyphen character is not accepted by the above regex, so probably an underscore should be used instead.

### Contracts

> A contract is the complete set of methods that an app implements. It

Reviewer 2: Why lowercase? Capitalized elsewhere and I like it as a proper noun

Author: Yes, there is a bit of inconsistency, because "interface" was always lower case. So not sure what convention to adopt.

Reviewer: I'd support capitalizing Interface and Contract everywhere for consistency

Author: I've modified this. Doing so, I've realized there may be some slightly confusing naming.
Let me know what you think. I see two types of objects:
* a "smart contract" or "application" or "app" or "stateful application" or "contract" is an actual application on Algorand
* a "Contract" is "the complete set of methods that an app implements"

Maybe to clarify should we restrict the vocabulary to:
* "application" for the one above
* "Contract" for the one below

so no "contract call transaction" but "application call transaction".

What do you think?

### Reference Types

Three special types are supported _only_ as the type of an

> argument. They **MUST** not be embedded in arrays or tuples, as 

Reviewer 2: MUST NOT

Author: Don't hesitate to use the suggestion button (leftmost button of the Github editor comment), which allows easier merging.
See:
```suggestion
argument. They **MUST NOT** not be embedded in arrays or tuples, as 
```

transactions are implied entirely by the placement of the transaction
types in the argument list.

For example, to invoke the method `deposit(string,axfer,pay,uint32)void`,
a client would create a transaction group containing, in this order:
1. an asset transfer
2. a payment

> 3. the actual application call
> When encoding the other (non-transaction) arguments, the client 

Reviewer: ```suggestion
3. the actual application call

When encoding the other (non-transaction) arguments, the client 
```

This needs another line break, otherwise the "When encoding..." sentence is part of the list

**MUST** act as if the transaction arguments were completely absent 
from the method signature. The application call would contain the method 
selector in ApplicationArgs[0], the first (string) argument in 
ApplicationArgs[1], and the fourth (uint32) argument in ApplicationArgs[2].

ARC-4 Apps **SHOULD** be constructed to allow their invocations to be
combined with other contract invocations in a single atomic group if

> they can do so safely. For example, they **SHOULD** use `txns` to examine
> the previous index in the group for a required `pay` transaction,
> rather than hardcode an index with `txn`.

Reviewer: ```suggestion
they can do so safely. For example, they **SHOULD** use `gtxns` to examine
the previous index in the group for a required `pay` transaction,
rather than hardcode an index with `gtxn`.
```

Tangential to this change, but this should be `gtxns` and `gtxn`

In general, an ARC-4 app method with n transactions as arguments **SHOULD** 
only inspect the n previous transactions. In particular, it **SHOULD NOT**
inspect transactions after and it **SHOULD NOT** check the size of a transaction 
group (if this can be done safely). 
In addition, a given method **SHOULD** always expect the same
number of transactions before itself. For example, the method 

> `deposit(string,axfer,pay,uint32)` is always preceded by two transactions.

Reviewer: ```suggestion
`deposit(string,axfer,pay,uint32)void` is always preceded by two transactions.
```
Return type missing
